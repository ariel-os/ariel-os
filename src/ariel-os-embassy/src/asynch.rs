//! Provides async functionality.
//!
//! This module bundles Ariel OS async functionality. It also wraps
//! some [Embassy](https://github.com/embassy-rs/embassy) types.

#![deny(missing_docs)]

use core::cell::OnceCell;

use embassy_sync::blocking_mutex::CriticalSectionMutex;

#[cfg(feature = "threading")]
pub mod blocker;

pub use embassy_executor::{SendSpawner, Spawner};
pub use embassy_futures::yield_now;

#[cfg(feature = "executor-thread")]
pub use crate::thread_executor;

pub(crate) static SPAWNER: CriticalSectionMutex<OnceCell<SendSpawner>> =
    CriticalSectionMutex::new(OnceCell::new());

/// Gets a spawner for the system executor.
///
/// # Panics
///
/// Panics when called before the system has finished initializing.
pub fn spawner() -> SendSpawner {
    SPAWNER.lock(|x| *x.get().unwrap())
}

/// Sets what `spawner()` returns.
///
/// May only be called once. Basically only in `super::init_task()`. That's why
/// we get away with ignoring the result.
#[allow(dead_code, reason = "actually used in `crate::init_task()`")]
pub(crate) fn set_spawner(spawner: SendSpawner) {
    let _ = SPAWNER.lock(|x| x.set(spawner));
}
