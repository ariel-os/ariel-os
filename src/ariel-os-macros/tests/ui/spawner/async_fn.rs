#![no_main]

// As the macro will fail, this import will not get used
#[allow(unused_imports)]
use ariel_os::asynch::Spawner;

// FAIL: spawner functions cannot be async
#[ariel_os::spawner(autostart)]
async fn main(spawner: Spawner) {}
