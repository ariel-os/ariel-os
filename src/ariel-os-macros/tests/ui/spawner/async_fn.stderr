error: custom attribute panicked
 --> tests/ui/spawner/async_fn.rs:8:1
  |
8 | #[ariel_os::spawner(autostart)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: message: spawner functions cannot be async, consider using `task` instead
